src/composition.rs

```diff
@@ -162,28 +162,68 @@ macro_rules! binary_eq_expr {
     };
 }
 
-fn logical_expr(
+#[derive(Clone, Copy)]
+struct FilterKey<T>(T);
+
+trait Get {
+    fn get(&mut self) -> Option<String>;
+}
+
+impl Get for FilterKey<&String> {
+    fn get(&mut self) -> Option<String> {
+        Some(self.0.to_owned())
+    }
```

Comment:
`FilterKey` is an abstraction layer for the key used in the filter clause.

For `add_filter()`, the key is a single string (a.k.a. "_field"):
`|> filter(fn: (r) => r._field == "firstField" or r._field == "secondField")`

For the `add_tag_values()`, the key is an array of strings:
`|> filter(fn: (r) => r.tagKey1 == "tagValue1" and r.tagKey2 == "tagValue2")`

By adding the FilterKey abstraction layer, we can reuse all the same infrastructure. The macro filter! and the fn logical_expr both still work, with minimal changes to support the abstraction type.

Reply:
@rockstar -- do I need to make this^^ into a code comment too?

Reply:
removed this code, per request